src/group_theory/submonoid/operations.lean

```diff
@@ -93,13 +93,86 @@ def add_submonoid.of_submonoid {M : Type*} [add_zero_class M] (S : submonoid (mu
+  (S.to_submonoid : set (multiplicative M)) = multiplicative.of_add '' S :=
+eq.symm (set.image_id S)
+
+lemma submonoid.of_add_submonoid_coe {M : Type*} [mul_one_class M]
+  (S : add_submonoid (additive M)) :
+  (submonoid.of_add_submonoid S : set M) = additive.to_mul '' S :=
+eq.symm (set.image_id S)
+
+lemma add_submonoid.of_submonoid_coe {M : Type*} [add_zero_class M]
+  (S : submonoid (multiplicative M)) :
+  (add_submonoid.of_submonoid S : set M) = multiplicative.to_add '' S :=
+eq.symm (set.image_id S)
+
 /-- Submonoids of monoid `M` are isomorphic to additive submonoids of `additive M`. -/
 def submonoid.add_submonoid_equiv (M : Type*) [mul_one_class M] :
-  submonoid M ≃ add_submonoid (additive M) :=
+  submonoid M ≃o add_submonoid (additive M) :=
 { to_fun := submonoid.to_add_submonoid,
   inv_fun := submonoid.of_add_submonoid,
   left_inv := λ x, by cases x; refl,
-  right_inv := λ x, by cases x; refl }
+  right_inv := λ x, by cases x; refl,
+  map_rel_iff' := λ a b, iff.rfl, }
+
+/-- Additive submonoids of monoid `M` are isomorphic to multiplicative submonoids of
+  `multiplicative M`. -/
+def add_submonoid.submonoid_equiv (M : Type*) [add_zero_class M] :
+  add_submonoid M ≃o submonoid (multiplicative M) :=
+{ to_fun := add_submonoid.to_submonoid,
+  inv_fun := add_submonoid.of_submonoid,
+  left_inv := λ x, by cases x; refl,
+  right_inv := λ x, by cases x; refl,
+  map_rel_iff' := λ a b, iff.rfl, }
+
+lemma submonoid.add_submonoid_equiv_coe (M : Type*) [add_zero_class M] :
+  ⇑(add_submonoid.submonoid_equiv M) = add_submonoid.to_submonoid := rfl
+
+lemma add_submonoid.submonoid_equiv_symm_coe (M : Type*) [add_zero_class M] :
+  ⇑(add_submonoid.submonoid_equiv M).symm = add_submonoid.of_submonoid := rfl
+
+lemma add_submonoid.submonoid_equiv_coe (M : Type*) [mul_one_class M] :
+  ⇑(submonoid.add_submonoid_equiv M) = submonoid.to_add_submonoid := rfl
+
+lemma submonoid.add_submonoid_equiv_symm_coe (M : Type*) [mul_one_class M] :
+  ⇑(submonoid.add_submonoid_equiv M).symm = submonoid.of_add_submonoid := rfl
+
+lemma submonoid.to_add_submonoid_mono {M : Type*} [mul_one_class M] :
+  monotone (submonoid.to_add_submonoid : submonoid M → add_submonoid (additive M)) :=
+λ a b hab, hab
+
+lemma add_submonoid.to_submonoid_mono {M : Type*} [add_zero_class M] :
+  monotone (add_submonoid.to_submonoid : add_submonoid M → submonoid (multiplicative M)) :=
+λ a b hab, hab
+
+lemma submonoid.closure.to_add_submonoid {M : Type*} [monoid M] (S : set M) :
+  (submonoid.closure S).to_add_submonoid = add_submonoid.closure (additive.of_mul '' S) :=
+begin
+  apply le_antisymm,
+  { apply (submonoid.add_submonoid_equiv M).to_galois_connection.l_le _,
+    rw [submonoid.closure_le, submonoid.add_submonoid_equiv_symm_coe,
+      submonoid.of_add_submonoid_coe, equiv.subset_image],
+    exact add_submonoid.subset_closure },
+  { rw [add_submonoid.closure_le, submonoid.to_add_submonoid_coe, equiv.image_subset],
+    exact submonoid.subset_closure }
+end
```

Comment:
The preimage version of the lemma is much easier to prove:
```suggestion
lemma submonoid.closure.to_add_submonoid {M : Type*} [monoid M] (S : set M) :
  (submonoid.closure S).to_add_submonoid = add_submonoid.closure (additive.to_mul ⁻¹' S) :=
le_antisymm
  ((submonoid.add_submonoid_equiv' M).to_galois_connection.l_le $
    submonoid.closure_le.2 add_submonoid.subset_closure)
  (add_submonoid.closure_le.2 submonoid.subset_closure)
```
I recommend changing all of the `'' S` in this PR to `⁻¹' S`

Reply:
Ah sorry, I completely misunderstood your suggestion, I tried to use `⁻¹'` on the left instead also changing `to` to `of`! Now it works, thanks!